### PR TITLE
8292006: Move thread accessor classes to threadJavaClasses.hpp

### PR DESCRIPTION
**Pull request by iklam**
To improve modularity and build time, move the declaration of the following accessor from classfile/javaClasses.hpp to runtime/threadJavaClasses.hpp:

+ java_lang_Thread_FieldHolder
+ java_lang_Thread_Constants
+ java_lang_ThreadGroup
+ java_lang_VirtualThread

Also move javaThreadStatus.hpp from share/classfile to share/runtime, where it belongs.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292006](https://bugs.openjdk.org/browse/JDK-8292006): Move thread accessor classes to threadJavaClasses.hpp

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9788/head:pull/9788` \
`$ git checkout pull/9788`

Update a local copy of the PR: \
`$ git checkout pull/9788` \
`$ git pull https://git.openjdk.org/jdk pull/9788/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9788`

View PR using the GUI difftool: \
`$ git pr show -t 9788`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9788.diff">https://git.openjdk.org/jdk/pull/9788.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/threadJavaClasses.inline.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+#ifndef SHARE_RUNTIME_THREADJAVACLASSES_INLINE_HPP
+#define SHARE_RUNTIME_THREADJAVACLASSES_INLINE_HPP
+
+#include "runtime/threadJavaClasses.hpp"
+
+#include "oops/instanceKlass.inline.hpp"
+#include "oops/oop.inline.hpp"
+#include "oops/oopsHierarchy.hpp"
```
Comment (tstuefe) [inline src/hotspot/share/runtime/threadJavaClasses.inline.hpp:32]: I believe that for INCLUDE_JFR you need macros.hpp, or? Highly likely it's already there, or you'd get linker errors. But it should be explicit.

File: src/hotspot/share/runtime/threadJavaClasses.hpp [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+ *
+ */
+
+#ifndef SHARE_RUNTIME_THREADJAVACLASSES_HPP
+#define SHARE_RUNTIME_THREADJAVACLASSES_HPP
+
+#include "classfile/vmClasses.hpp"
+#include "memory/allStatic.hpp"
+#include "oops/oopsHierarchy.hpp"
+#include "runtime/os.hpp"
```
Comment (tstuefe) [inline src/hotspot/share/runtime/threadJavaClasses.hpp:31]: Where do you need os.hpp?

File: src/hotspot/share/runtime/threadJavaClasses.cpp [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+#include "classfile/javaClassesImpl.hpp"
+#include "classfile/vmSymbols.hpp"
+#include "memory/oopFactory.hpp"
+#include "runtime/continuationEntry.inline.hpp"
+#include "runtime/continuationJavaClasses.inline.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "runtime/javaThreadStatus.hpp"
+#include "runtime/threadJavaClasses.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/vframe.inline.hpp"
```
Comment (tstuefe) [inline src/hotspot/share/runtime/threadJavaClasses.cpp:36]: needs debug.hpp and macros.hpp, at least, to be self-contained. Probably globalDefinitions.hpp too.

**Review by tstuefe — APPROVED** [review]
Hi Ioi,

looks good. Small nits inline. I did not check that the moved code is identical, because I assume you just moved it around unchanged.

Cheers, Thomas

**Comment by dholmes-ora** [comment]
<grumble grumble> javaClasses is not meant to be modular, and moving just a handful of items does not make it modular. I really like having all these Java accessors in one place.

File: src/hotspot/share/runtime/threadJavaClasses.inline.hpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/threadJavaClasses.inline.hpp:32]: It is included in the .hpp file as it should be.

**Comment by dholmes-ora** [comment]
My issue with these kinds of changes in general is that they:
- undermine institutional memory
- make it harder to do backports due to the shuffling
- make it harder to track changes due to the shuffling

So the benefit to me has to be more than just a small decrease in some build times which most people are never going to notice anyway. Sorry. Just my 2c.

**Comment by iklam** [comment]
> My issue with these kinds of changes in general is that they:
> 
> * undermine institutional memory
> * make it harder to do backports due to the shuffling
> * make it harder to track changes due to the shuffling
> 
> So the benefit to me has to be more than just a small decrease in some build times which most people are never going to notice anyway. Sorry. Just my 2c.

While I understand the concerns, the above reason can be used against any sort of refactoring of the source code. We have a few decades of development, resulting slow but steady source code rot. We need to refactor the code when it makes sense to keep the repo healthy. Yes, the above are valid concerns, but we have been successful in dealing with them in past refactoring exercises.

Build time is important for HotSpot developers. I figure that the header file refactoring that happened in the last year or so (by myself and others) have resulted in 20% or so reduction in build time. So of that was taken back, though, due to new features such as Loom.

I tried to limit the amount of code to be moved. However, when a file like javaClasses.hpp grows to over 2000 lines or mostly unrelated declarations, it's a sign that it needs to be broken up (which I started doing since JDK-8288623).

**Comment by dholmes-ora** [comment]
> the above reason can be used against any sort of refactoring of the source code

Indeed - they are always the "cons" of such a refactoring, and it is up to the "pros" to outweigh them.

**Pull request closed without merging**